src/compiler/checker.ts

```diff
@@ -8831,7 +8833,18 @@ export function createTypeChecker(host: TypeCheckerHost): TypeChecker {
                     else {
                         const type = getTypeOfSymbol(symbol);
                         const localName = getInternalSymbolName(symbol, symbolName);
-                        if (!(symbol.flags & SymbolFlags.Function) && isTypeRepresentableAsFunctionNamespaceMerge(type, symbol)) {
+                        if (type.symbol && type.symbol !== symbol && type.symbol.flags & SymbolFlags.Function && some(type.symbol.declarations, isFunctionExpressionOrArrowFunction) && (type.symbol.members?.size || type.symbol.exports?.size)) {
+                            // assignment of a anonymous expando/class-like function, the func/ns/merge branch below won't trigger,
+                            // and the assignment form has to reference the unreachable anonymous type so will error.
+                            // Instead, serialzie the type's symbol, but with the current symbol's name, rather than the anonymous one.
+                            if (!context.remappedSymbolRefernces) {
+                                context.remappedSymbolRefernces = new Map();
+                            }
+                            context.remappedSymbolRefernces.set(getSymbolId(type.symbol), symbol); // save name remapping as local name for target symbol

```

Comment:
This is to handle the not-anonymous-but-still-an-expression case where you could use that local name to refer to the type of the expando-thing as a whole. Because it'd be odd if just adding a name to the function expression made this transform stop working.

Reply:
Is this is for cases like

```js
const f = function g() { }
f.p = 1
```

Does remappedSymbolReferences record `g`? When would `g` show up in the declaration emit?

Edit: Looking at the test cases, it's actually `g -> f` so that when a constructor function `g` returns `g`, it serialises as `f`.